plugins/logitech-scribe/fu-logitech-scribe-device.c

```diff
@@ -0,0 +1,603 @@
+				   GError **error)
+{
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	chunks = fu_chunk_array_new_from_bytes(fw, 0x0, 0x0, PAYLOAD_SIZE);
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+		g_autoptr(GByteArray) data_pkt = g_byte_array_new();
+		g_byte_array_append(data_pkt, fu_chunk_get_data(chk), fu_chunk_get_data_sz(chk));
+		if (!fu_logitech_scribe_device_send_upd_cmd(self,
+							    CMD_DATA_TRANSFER,
+							    data_pkt,
+							    error)) {
+			g_prefix_error(error, "failed to send data packet 0x%x: ", i);
+			return FALSE;
+		}
+		fu_progress_step_done(progress);
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_logitech_scribe_device_write_firmware(FuDevice *device,
+					 FuFirmware *firmware,
+					 FuProgress *progress,
+					 FwupdInstallFlags flags,
+					 GError **error)
+{
+	FuLogitechScribeDevice *self = FU_LOGITECH_SCRIBE_DEVICE(device);
+	g_autofree gchar *base64hash = NULL;
+	g_autoptr(GByteArray) end_pkt = g_byte_array_new();
+	g_autoptr(GByteArray) start_pkt = g_byte_array_new();
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "init");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, "start-transfer");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 94, "device-write-blocks");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 1, "end-transfer");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 5, "uninit");
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* sending INIT. Retry if device is not in IDLE state to receive the file */
+	if (!fu_device_retry(device,
+			     fu_logitech_scribe_device_send_upd_init_cmd_cb,
+			     MAX_RETRIES,
+			     NULL,
+			     error)) {
+		g_prefix_error(error,
+			       "failed to write init transfer packet: please reboot the device: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* start transfer */
+	fu_byte_array_append_uint64(start_pkt, g_bytes_get_size(fw), G_LITTLE_ENDIAN);
+	if (!fu_logitech_scribe_device_send_upd_cmd(self, CMD_START_TRANSFER, start_pkt, error)) {
+		g_prefix_error(error, "failed to write start transfer packet: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* push each block to device */
+	if (!fu_logitech_scribe_device_write_fw(self, fw, fu_progress_get_child(progress), error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* end transfer */
+	base64hash = fu_logitech_scribe_device_compute_hash(fw);
+	fu_byte_array_append_uint32(end_pkt, 1, G_LITTLE_ENDIAN);   /* update */
+	fu_byte_array_append_uint32(end_pkt, 0, G_LITTLE_ENDIAN);   /* force */
+	fu_byte_array_append_uint32(end_pkt, MD5, G_LITTLE_ENDIAN); /* checksum type */
```

Comment:
I think this needs to be FU_LOGITECH_SCRIBE_CHECKSUM_KIND_MD5 or something. Using such generic #defines might lead to trouble.